### PR TITLE
Move error formatting functionality 

### PR DESCRIPTION
**Pull request by scoopr**
**Connections**
This is an attempt to implement #1082 

**Description**
This is a fairly straightforward move of the error formatting that were in wgpu-rs in to wgpu-code.

The enriching of some of the errors with the additional `ContextError` were still left with the wgpu-rs code, as I realised the current error messages refer to the API names as they appear in rust code and might not make sense for all the consumers of wgpu-core. It could be left as is and other consumers might enrich in their own ways, or the messages could be reformatted to be more conversational

**Testing**
Limited testing with the examples

### CODE REVIEW

**Comment by scoopr** [comment]
It may be a bit rough around the edges still.

As it is, I don't it really bring much progress/clarity/etc to the code, but I think the biggest reason was to get the gecko error formatting to use the same formatting mechanism eventually.

Aside, "bringing the family back together" lowered the bar significantly for me to even start with this :)

File: wgpu-core/src/error.rs
Comment (kvark) [inline wgpu-core/src/error.rs:68]: Would it make sense to move each implementation into the module that own the type (i.e. `CreatePipelineLayoutError` in this case)? This way, if the error gets more variants, it's easier to see the associated code.

File: wgpu-core/src/error.rs
Comment (kvark) [inline wgpu-core/src/error.rs:250]: It would be elegant to make it accept an `fmt::Formatter`, so that it doesn't allocate anything.
Yes, I know the performance doesn't matter here, so this isn't required to be elegant :)

File: wgpu-core/src/error.rs
Comment (kvark) [inline wgpu-core/src/error.rs:9]: Looks like this is only needed for just a few cases. Could it be worked around?
Something like https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ced1f8a2ddc2c0b986421b1acd195588

File: wgpu-core/src/error.rs
Comment (kvark) [inline wgpu-core/src/error.rs:133]: this should probably look like `format_label_line(&mut ret, "bind group", &name)`, so that it writes to the string directly instead of first creating a new string and then taking a reference of it

**Review by kvark — COMMENTED** [review]
Very nice!

File: wgpu-core/src/error.rs
Comment (scoopr) [reply wgpu-core/src/error.rs:68]: Ah yeah, true, now that that is possible and makes sense.
Does the `PrettyError` name still make sense if it is used more widely?

I'm somehow still hoping for a way to implement the formatting with the standard `Display` trait, but I guess there is still no way of wiggling the Global in there to resolve the labels. Does pythonesques rewrite affect this, would the label be accessible just from the Id? That would simplify the whole thing, if we could just format the errors like any other errors.

File: wgpu-core/src/error.rs
Comment (scoopr) [reply wgpu-core/src/error.rs:250]: I think I attempted that while back when I was initially implementing this, and I couldn't figure it out quite right.
The problem is that you can't really construct a `fmt::Formatter`, so it is only really useful in a context where you already have it, like within `Display`/`Debug`. 

Current call sites want the result as a string, and as I understood it, I would have to wrap the stuff to something that implements Display to get the result, and I didn't want to deal with that.

I suppose it could return `fmt::Arguments`, but I guess it would do some allocation anyway? 

If we happen to get rid of the whole `PrettyError` in favour of the `Display`, then the issue is moot and `fmt::Formatter` is the easiest choice.

File: wgpu-core/src/error.rs
Comment (scoopr) [reply wgpu-core/src/error.rs:9]: Yeah, I don't remember the whole thought process why it originally was like that, but that looks like workable solution.

File: wgpu-core/src/error.rs
Comment (scoopr) [reply wgpu-core/src/error.rs:133]: Hmm yeah, if the `fmt::Formatter` etc. changes don't pan out, then that seems like a worthwhile change.

File: wgpu-core/src/error.rs
Comment (scoopr) [reply wgpu-core/src/error.rs:9]: Hmm, couldn't get it quite to work.
```
pub fn format_pretty_any(
    global: &Global<IdentityManagerFactory>,
    error: &(dyn Error + 'static),
) -> String
```
In here, even though `Error´ requires Display, somehow it won't allow casting to it. Adding the trait to the trait object requirement doesn't work and suggests making a wrapper trait that requires the two things, but `Error` already requires `Display`, so that seems like a dead-end.

```
pub trait PrettyError: Error {
    fn fmt_pretty(&self, _global: &Global<IdentityManagerFactory>) -> String {
        let self_display = self as &dyn fmt::Display;
        format_error_line(self_display)
    }
}
```
In here it says self doesn't have compile time size, though I'm not entirely sure why it requires that, isn't trait object a compile time known size?

Either I'm missing some syntax that I'm forgetting, or its not really possible to do it like that.

File: wgpu-core/src/error.rs
Comment (kvark) [reply wgpu-core/src/error.rs:250]: sorry, it would be generic over `fmt::Write`, not `fmt::Formatter`

Comment (scoopr) [reply wgpu-core/src/error.rs]: Oh derp. I should of realized that myself, I implemented this now, it somehow makes a bit more sense now :)

**Comment by scoopr** [comment]
In the latest wip commit, I attempted to move the PrettyError impls closer to the error itself.

As I was trying it out in the naive way of just moving the code, I hated how all kinds of types started to leak out everywhere (PrettyError, Global, gfx_select, etc.), so I thought it could use some abstraction over it.

So in here, I tried a `PrettyError2`, just for `RenderCommandError`, to feel it out. I end up throwing the Write and Global refs in to this wrapper struct, that has convenience methods for formatting the lines, I even ended up making a convenience method per label type, which nicely wraps over the gfx_select stuff. The using code doesn't really need to know more than the `PrettyError2` type itself to be useful.

Do you think this is an approach worth pursuing, or is it too heavy handed?

File: wgpu-core/src/error.rs
Comment (kvark) [reply wgpu-core/src/error.rs:106]: cc @pythonesque

> Does the PrettyError name still make sense if it is used more widely?

seems fine, unless you have ideas for a better name ;)

**Review by kvark — COMMENTED** [review]
Not seeing anything wrong with this approach. If you have the set of trade-offs in mind for this, please list them up. Maybe that will ring a bell for me :)

**Comment by pythonesque** [comment]
Seems fine to me :)  All the methods that take Hubs/Global (that are currently needed to extract labels) are just static methods with hubs removed, so this change should extend logically to a Hubs-free world.

**Comment by scoopr** [comment]
If the Global is truly going away with the refactor, I’m somewhat tempted to hold off until that lands, as this could just be a “simply” a Display impl (the error contains everything it needs to display itself). 

Thou I suppose it would be odd to have the Display show the notes/labels, especially if it assumes some form of formatting/indent etc.

What I’m kind of aiming at, is for the user code to be able to format error messages with the standard Error mechanism as it sees fit, but with enough of granular access to the labels etc. to be able to do some rudimentary colouring of the message. Some more advanced cases might try to even explain the source data where the label came from, if it can access it when available. Buuut I might be overthinking it :)

Maybe I’ll just continue with the trait, but impl the Display with it, and see where it takes me to.

**Comment by scoopr** [comment]
Alright, here is the current result. PrettyError impls are now next to the error enums, the formatting is done through the ErrorFormatter thing introduced there.

I also worked the AsDisplay hack, so it gets removed too. I suppose if you have a trait object already, you can't pass it as another trait object, even if its a traits required trait, as far as I can see. Also some magic Sized in there

File: wgpu-core/src/error.rs
Comment (kvark) [inline wgpu-core/src/error.rs:20]: any reason to not use "\t" here?

File: wgpu-core/src/error.rs
Comment (kvark) [inline wgpu-core/src/error.rs:19]: can this work with `FormatArgs`?

**Review by kvark — COMMENTED** [review]
Nice work!

File: wgpu-core/src/error.rs
Comment (scoopr) [reply wgpu-core/src/error.rs:20]: Are we going to have _that_ discussion, tabs vs spaces? ;)
I didn't really intend functional changes here, and that's how it was. I feel the tabs would indent a bit too much though.

For reference, `anyhow` seems to be doing [indenting with spaces](https://github.com/dtolnay/anyhow/blob/master/src/fmt.rs#L73)

File: wgpu-core/src/error.rs
Comment (scoopr) [reply wgpu-core/src/error.rs:19]: Do you mean `fmt::Arguments`? And that the note/error/label stuff would return it instead of writing directly to the `Writer`? I don't really see the point, the users (PrettyError impls) would just pass it on directly to a `Writer` anyway?

Or do you mean instead of taking a `&dyn fmt::Display` argument, take a `&fmt::Arguments`? I suppose thats ok, though the caller would need to do some extra `format_args!`, that is not needed now?

Comment (scoopr) [reply wgpu-core/src/error.rs:19]: Ooh, do you mean to avoid the subsequent formatting, collapsing them to a single one, by passing fmt::Arguments until the note thing that writes it out?

Comment (kvark) [reply wgpu-core/src/error.rs:19]: yeah, but you are right, it doesn't matter too much

**Review by kvark — APPROVED** [review]

**Pull request merged**
